[DI][Config] Add & use ReflectionClassResource

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #21079
| License       | MIT
| Doc PR        | -

With new changes comming to 3.3, we need a more generic reflection tracking logic than the one already managed by the autowiring subsystem.

This PR adds a new ReflectionClassResource in the Config component, and a new ContainerBuilder::getReflectionClass() method in the DI one (for fetching+tracking reflection-related info).

ReflectionClassResource tracks changes to any public or protected properties/method.

PR updated and ready, best viewed [ignoring whitespaces](https://github.com/symfony/symfony/pull/21419/files?w=1).

changelog:

* added `ReflectionClassResource` class
* added second `$exists` constructor argument to `ClassExistenceResource` - with a special mode that prevents fatal errors from happening when some parent class is broken (logic generalized from AutowiringPass)
* made `ClassExistenceResource` also work with interfaces and traits
* added `ContainerBuilder::getReflectionClass()` for retrieving and tracking reflection class info
* deprecated `ContainerBuilder::getClassResource()`, use `ContainerBuilder::getReflectionClass()` or `ContainerBuilder::addObjectResource()` instead

nicolas-grekas committed:
[DI][Config] Add & use ReflectionClassResource

stof commented on src/Symfony/Component/Config/Tests/Resource/ReflectionClassResourceTest.php:
```diff
@@ -0,0 +1,57 @@
+
+        unlink($tmp);
+        $this->assertFalse($res->isFresh($now), '->isFresh() returns false if the resource does not exist');
+    }
```
is there a way to test the handling of the hash ?

nicolas-grekas replied:
I don't see how for now :(

nicolas-grekas replied:
I split the logic to allow for testing signature mutations - tests added

stof commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -459,6 +424,9 @@ private function addServiceToAmbiguousType($id, $type)
     }
 
+    /**
+     * @deprecated since version 3.3, to be removed in 4.0.
```
deprecating a private method does not make sense. Either we don't use it anymore, and we should remove it already, or we use it in deprecated method and it will become dead code when removing the other method (which is catched by IDEs and any static analysis tool, for instance SensioLabsInsight)

nicolas-grekas replied:
It's just a friendly reminder for the one who will cleanup the code preparing 4.0

stof commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -279,6 +285,36 @@ public function addClassResource(\ReflectionClass $class)
+     *
+     * @param string $class
+     *
+     * @return \ReflectionClass|false
```
Can we use ``\ReflectionClass|null`` instead ? It would be cleaner (PHP 7.1 has nullable types, not falsable types)

nicolas-grekas replied:
would only add boilerplate (`isset(false)` is used in the code)

stof replied:
Well, if this was a private API, I would be OK with this argument about simplifying the internal implementation.
But this is a public API.

fabpot replied:
@nicolas-grekas I think we need to try to not use mixed return args in our public APIs (anymore). With `\ReflectionClass|null`, we could add a native PHP return type in 7.1, but `\ReflectionClass|false`, we cannot.

nicolas-grekas replied:
done

stof commented on src/Symfony/Component/DependencyInjection/composer.json:
```diff
@@ -20,7 +20,7 @@
     "require-dev": {
         "symfony/yaml": "~3.2",
-        "symfony/config": "~2.8|~3.0",
+        "symfony/config": "~3.3",
```
should we add a conflict rule to avoid issues when getting a lower version for people using standalone components ?

nicolas-grekas replied:
good idea, added

dunglas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -56,9 +55,13 @@ public function process(ContainerBuilder $container)
      */
     public static function createResourceForClass(\ReflectionClass $reflectionClass)
     {
+        @trigger_error('The '.__METHOD__.'() method is deprecated since version 3.3 and will be removed in 4.0. Use ContainerBuilder::getReflectionClass() instead.', E_USER_DEPRECATED);
```
Not sure if it's a good idea or not. The current implementation will more efficient because some changes doesn't impact autowiring (like properties changes).

nicolas-grekas replied:
In practice, I don't think there are that more situations where the code would not-invalidate where the existing would. And I see the benefit for having only one place where this tracking logic is implemented. So I think it's in fact a desirable thing yes. Allows better aggregation of resources across passes also - and maybe faster checking (with simpler/optimized logic.)

dunglas commented:
👍

nicolas-grekas commented:
|null done

nicolas-grekas commented:
PR updated and ready, best viewed [ignoring whitespaces](https://github.com/symfony/symfony/pull/21419/files?w=1).

changelog:

* added `ReflectionClassResource` class
* added second `$exists` constructor argument to `ClassExistenceResource` - with a special mode that prevents fatal errors from happening when some parent class is broken (logic generalized from AutowiringPass)
* made `ClassExistenceResource` also work with interfaces and traits
* added `ContainerBuilder::getClassExists()` for checking and tracking class, interface or trait existence
* added `ContainerBuilder::getReflectionClass()` for retrieving and tracking reflection class info
* deprecated `ContainerBuilder::getClassResource()`, use `ContainerBuilder::getReflectionClass()` or `ContainerBuilder::addObjectResource()` instead

nicolas-grekas commented:
addObjectResource and ReflectionClassResource now track mtimes of parents, traits and interfaces.

Hash computation, file and class existence is now done lazily - ie after the resources deduplication.

And to prevent checking the hash again when it was found valid in a previous run, files are touched when needed.

nicolas-grekas commented:
To be merged on top of #21452 - which holds the first commit of this PR for now.

nicolas-grekas commented:
Rebased
ping @stof :)

stof commented on src/Symfony/Component/Config/Resource/ReflectionClassResource.php:
```diff
@@ -0,0 +1,185 @@
+        yield $class->getDocComment().$class->getModifiers();
+
+        if ($class->isTrait()) {
+            yield print_r(class_uses($class->name), true);
```
shouldn't this be always included ?

nicolas-grekas replied:
no need: traits are not directly part of the signature - the method they provide will be seen as any other methods, thus tracked

stof commented on src/Symfony/Component/Config/Tests/Resource/ClassExistenceResourceTest.php:
```diff
@@ -51,4 +51,20 @@ public function testIsFreshWhenClassExists()
+
+    public function testExistsKo()
+    {
+        spl_autoload_register(function ($class) use (&$loadedClass) { $loadedClass = $class; });
```
you should unregister it at the end of the test (even when it fails)

stof commented on src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php:
```diff
@@ -30,13 +32,17 @@ public function process(ContainerBuilder $container)
+
+            if (!$r = $container->getReflectionClass($class)) {
+                throw new InvalidArgumentException(sprintf('Class "%s" used for service "%s" cannot be found.', $class, $id));
+            } elseif (!$r->isSubclassOf(Command::class)) {
```
just ``if``

stof commented on src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php:
```diff
@@ -30,13 +32,17 @@ public function process(ContainerBuilder $container)
 
             if ($definition->isAbstract()) {
-                throw new \InvalidArgumentException(sprintf('The service "%s" tagged "console.command" must not be abstract.', $id));
+                continue;
```
why changing this ?

stof commented on src/Symfony/Component/Config/Resource/ReflectionClassResource.php:
```diff
@@ -0,0 +1,185 @@
+                // ensure the next checks won't require a hash computation again,
+                // use a local mtime cache so that the current process still sees
+                // the original mtime and validates the hash for child classes
+                @touch($file, $timestamp);
```
this might break other systems relying on the real mtime of the file because they are concerned about more things than the ones involved in the Symfony hash

nicolas-grekas replied:
true, removed

stof commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -251,7 +253,32 @@ public function setResources(array $resources)
     {
         if ($this->trackResources) {
-            $this->addClassResource(new \ReflectionClass($object));
+            if (is_object($object)) {
```
if it also accepts class names, the phpdoc must be updated

stof commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -263,20 +290,57 @@ public function addObjectResource($object)
     {
-        if (!$this->trackResources) {
-            return $this;
+        @trigger_error('The '.__METHOD__.'() mehtod is deprecated since version 3.3 and will be removed in 4.0. Use the addObjectResource() or the getReflectionClass() method instead.', E_USER_DEPRECATED);
```
typo

stof commented on src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php:
```diff
@@ -94,17 +95,13 @@ public function process(ContainerBuilder $container)
-                }
-
-                throw new InvalidArgumentException(sprintf('Service "%s" must implement interface "%s".', $id, $interface));
+            if (!$r = $container->getReflectionClass($class)) {
```
We actually need to invalidate the cache when the implementation of the class changes here, as it can change the list of events (btw, it means we are missing a ClassResource today)

nicolas-grekas replied:
good catch, fixed

nicolas-grekas commented:
@stof comments addressed

stof commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -177,16 +175,6 @@ public function dump(array $options = array())
-        foreach ($this->classResources as $r) {
-            $this->container->addClassResource($r);
-        }
-        $this->classResources = array();
```
do we actually have this duplicate code in older branches, or was it all added in 3.3 only ?

nicolas-grekas replied:
3.3 only

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -263,20 +290,57 @@ public function addObjectResource($object)
+        if ($this->trackResources) {
+            if (!$classReflector) {
+                $this->addResource($resource ?: new ClassExistenceResource($class, ClassExistenceResource::EXISTS_KO));
+            } elseif (!$classReflector->isInternal()) {
```
We can track details for internal classes, but this makes no sense: if you switch your version/extension of PHP, the container is not invalidated. Tracking here would mean checking the signature every single time. Makes no sense to me.

nicolas-grekas replied:
See #21505 for internal classes

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1324,42 +1424,6 @@ public static function getServiceConditionals($value)
-     *
-     * @final
-     */
-    public function fileExists($path, $trackContents = true)
```
moved as is above with other resource-related methods

fabpot commented:
Thank you @nicolas-grekas.

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/composer.json:
```diff
@@ -22,7 +22,7 @@
         "symfony/config": "~3.3",
         "symfony/event-dispatcher": "~3.3",
-        "symfony/http-foundation": "~3.1",
+        "symfony/http-foundation": "~3.3",
```
What is the reason for this change?

nicolas-grekas replied:
Transitivity and composer bug, this doesn't change the result, it only helps composer

xabbuh replied:
thanks, I was just wondering as the code changes didn't indicate any need for this change
